modules/core/src/main/java/org/apache/ignite/internal/hlc/HybridClockImpl.java

```diff
@@ -111,20 +111,30 @@ public final HybridTimestamp current() {
      */
     @Override
     public final HybridTimestamp update(HybridTimestamp requestTime) {
-        while (true) {
-            long now = currentTime();
+        long requestTimeLong = requestTime.longValue();
 
+        while (true) {
             // Read the latest time after accessing UTC time to reduce contention.
-            long oldLatestTime = latestTime;
+            long oldLatestTime = this.latestTime;
 
-            long newLatestTime = max(requestTime.longValue() + 1, max(now, oldLatestTime + 1));
+            if (oldLatestTime >= requestTimeLong) {
+                return hybridTimestamp(LATEST_TIME.incrementAndGet(this));
+            }
 
-            // TODO https://issues.apache.org/jira/browse/IGNITE-23707 avoid CAS on logical part update.
+            long now = currentTime();
+
+            if (now > requestTimeLong) {
+                if (LATEST_TIME.compareAndSet(this, oldLatestTime, now)) {
+                    return hybridTimestamp(now);
+                }
+            } else {
+                long newLatestTime = requestTimeLong + 1;
 
-            if (LATEST_TIME.compareAndSet(this, oldLatestTime, newLatestTime)) {
-                notifyUpdateListeners(newLatestTime);
+                if (LATEST_TIME.compareAndSet(this, oldLatestTime, newLatestTime)) {
```

Comment:
Jira says that we should do an atomic increment in this branch:
> We can skip CAS if only logical part of clocks need to be updated.
This can be achieved using atomic increment, which is more efficient.

I still see a CAS here. Please replace it with `incrementAndGet`, or explain why can't do this.

Do we have any benchmark results here?

Reply:
Here, we have to replace an internal timestamp with the timestamp from the method parameter. It is impossible to achieve this by incrementing.

Reply:
> Do we have any benchmark results here?
No, I did not run a benchmark.
We already estimated that the incrementAndGet has a less impact, either on latency or throughput.

Reply:
incrementAndGet definitely works better under contention

Reply:
> Here, we have to replace an internal timestamp with the timestamp from the method parameter. It is impossible to achieve this by incrementing.

I see, my mistake, thank you for the reply!